HDDS-8573. Verify default setting for DN root dir to restrict non-admin access

ArafatKhan2198 opened the pull request:
## What changes were proposed in this pull request?

The recommendation to set the permissions for the Apache Ozone DN, OM, SCM, Reccon metadata storage directories to 750 or tighter means that only the owner of the directory (in this case, the Hadoop user) and members of the Hadoop group should have access to the directory. By setting the permissions to 750, non-root users will not be able to read or write to the storage directories, which is important for protecting sensitive data.

The table represents the permissions configuration associated with each configuration property :- 
| Configuration Property               | Permissions |
|--------------------------------------|-------------|
| ozone.recon.db.dirs.permissions      | 750         |
| ozone.scm.db.dirs.permissions        | 750         |
| ozone.om.db.dirs.permissions         | 750         |
| ozone.metadata.dirs.permissions      | 750         |

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8573

## How was this patch tested?
### Unit Tests
### Manually :- 
The final permissions for the **_metadata directory_** will be set according to the configured permissions. In our case, the default is 750, which corresponds to the permissions "rwxr-x---".
```
- Ozone Manager 
bash-4.2$ ls -al
total 12
drwxrwxrwt 1 root   root  4096 May 13 06:28 .
drwxr-xr-x 1 root   root  4096 May 13 06:28 ..
drwxr-x--- 7 hadoop users 4096 May 13 06:28 metadata

- Storage Container Manager 
sh-4.2$ ls -al
total 12
drwxrwxrwt  1 root   root  4096 May 13 06:28 .
drwxr-xr-x  1 root   root  4096 May 13 06:28 ..
drwxr-x--- 10 hadoop users 4096 May 13 06:28 metadata

- Recon 
sh-4.2$ ls -al 
total 12
drwxrwxrwt 1 root   root  4096 May 13 06:28 .
drwxr-xr-x 1 root   root  4096 May 13 06:28 ..
drwxr-x--- 8 hadoop users 4096 May 13 06:29 metadata

- DataNode 
sh-4.2$ ls -al
total 20
drwxrwxrwt 1 root   root  4096 May 13 06:28 .
drwxr-xr-x 1 root   root  4096 May 13 06:28 ..
-rw-r--r-- 1 hadoop users  357 May 13 06:28 datanode.id
drwxr-xr-x 3 hadoop users 4096 May 13 06:28 hdds
drwxr-x--- 6 hadoop users 4096 May 13 06:28 metadata
```

krishnaasawa1 commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/ScmConfigKeys.java:
```diff
@@ -212,6 +216,8 @@ public final class ScmConfigKeys {
       "hdds.rest.http-address";
   public static final String HDDS_REST_HTTP_ADDRESS_DEFAULT = "0.0.0.0:9880";
   public static final String HDDS_DATANODE_DIR_KEY = "hdds.datanode.dir";
+  public static final String HDDS_DATANODE_DIR_PERMISSION =
```
How this is changing default to 700 for DN storage dirs without specifing in ozone defaults?

krishnaasawa1 commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -683,6 +691,14 @@
   </property>
+  <property>
+    <name>ozone.metadata.dirs.permissions</name>
+    <value>777</value>
```
Do default needs to more restrictive 750 or 755 or what's advantage of adding this and other properties below with 777 or full permission?

ArafatKhan2198 commented:
@krishnaasawa1 @devmadhuu my apologies for mistakenly applying the wrong changes to the recent patch submission. I have now corrected the mistake and updated the PR with the intended changes, including `configurable configurations` and also added unit tests for validation. Please take a look.

adoroszlai commented:
@ArafatKhan2198 please take a look at failed tests:

```
org.apache.hadoop.ozone.recon.TestReconUtils
org.apache.hadoop.ozone.recon.api.TestBlocksEndPoint
org.apache.hadoop.ozone.recon.api.TestContainerEndpoint
org.apache.hadoop.ozone.recon.api.TestContainerStateCounts
org.apache.hadoop.ozone.recon.api.TestEndpoints
org.apache.hadoop.ozone.recon.api.TestNSSummaryEndpointWithFSO
org.apache.hadoop.ozone.recon.api.TestNSSummaryEndpointWithLegacy
org.apache.hadoop.ozone.recon.api.TestOpenContainerCount
org.apache.hadoop.ozone.recon.api.TestTriggerDBSyncEndpoint
org.apache.hadoop.ozone.recon.recovery.TestReconOmMetadataManagerImpl
org.apache.hadoop.ozone.recon.spi.impl.TestOzoneManagerServiceProviderImpl
org.apache.hadoop.ozone.recon.tasks.TestContainerKeyMapperTask
org.apache.hadoop.ozone.recon.tasks.TestNSSummaryTask
org.apache.hadoop.ozone.recon.tasks.TestNSSummaryTaskWithFSO
org.apache.hadoop.ozone.recon.tasks.TestNSSummaryTaskWithLegacy
org.apache.hadoop.ozone.recon.tasks.TestTableCountTask
```

https://github.com/ArafatKhan2198/ozone/actions/runs/4965532983/jobs/8886411048#step:6:8616

```
org.apache.hadoop.ozone.TestOzoneConfigurationFields
org.apache.hadoop.ozone.recon.TestReconAsPassiveScm
org.apache.hadoop.ozone.recon.TestReconScmHASnapshot
org.apache.hadoop.ozone.recon.TestReconScmSnapshot
org.apache.hadoop.ozone.recon.TestReconTasks
org.apache.hadoop.ozone.recon.TestReconWithOzoneManager
org.apache.hadoop.ozone.recon.TestReconWithOzoneManagerFSO
org.apache.hadoop.ozone.recon.TestReconWithOzoneManagerHA
org.apache.hadoop.ozone.shell.TestNSSummaryAdmin
```

https://github.com/ArafatKhan2198/ozone/actions/runs/4965532983/jobs/8886411244#step:5:4101

ashishkumar50 commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestServerUtils.java:
```diff
@@ -18,25 +18,145 @@
+    // Create an OzoneConfiguration object and set the permissions
+    // for different keys
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.set("ozone.recon.db.dir.perm", "750");
```
This needs to be corrected to "ozone.recon.db.dirs.permissions"

ArafatKhan2198 replied:
Thanks it's corrected !!

jojochuang replied:
Can you use the static variable OZONE_RECON_DB_DIRS_PERMISSIONS instead?
Replace the string literals below with static variables in the blow too.

ArafatKhan2198 replied:
Thanks for the comment @jojochuang I have made the changes!

ashishkumar50 commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -669,6 +669,14 @@
+    <value>750</value>
+    <description>
+      Permissions for the metadata directories for Ozone Manager The
+      permissions have to be octal.
```
As SCM,Recon DB directories you marked it can be either octal or symbolic, I think OM directory permission can also be symbolic and not only octal.

ArafatKhan2198 replied:
Thanks for the suggestion! Initially, I believed that providing octal permissions for configurations would be sufficient. However, since you mentioned the importance of supporting both octal and symbolic formats, I realized it makes sense. It will be easier for users to provide permissions in the symbolic format rather than octal. Therefore, I have made the necessary changes to support both formats and have also added unit tests to ensure everything works correctly. 

@ashishkumar50

ashishkumar50 reviewed (commented):
Hi @ArafatKhan2198, Thanks for working on this. Please find few comments.

jojochuang commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java:
```diff
@@ -178,12 +184,90 @@ public static File getDirectoryFromConfig(ConfigurationSource conf,
             dbDirPath + " specified in configuration setting " +
             key);
       }
+      try {
```
Can you update the javadoc of this method so that users of the method is aware it has side effects?

ArafatKhan2198 replied:
Sure will do!!

jojochuang commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/server/ServerUtils.java:
```diff
@@ -178,12 +184,90 @@ public static File getDirectoryFromConfig(ConfigurationSource conf,
+    }
+
+    throw new IllegalArgumentException(
+        "Invalid configuration value for key: " + key);
```
If configValue is null it means the value for the corresponding configuration key is not defined.

ArafatKhan2198 replied:
The get() method in the `ConfigurationSource` interface utilizes the in-memory map of configurations and their respective values stored in the `InMemoryConfiguration` class. This map holds the defined configurations. When the specified key does not exist in the map, the get() method will return null. Similarly, if the key has not been set or is not present in the map, the method will not find a corresponding value and also return null.

Hence, in a case where the configuration does not exist which means a null value will be given to `configValue`, an `IllegalArgumentException` will be thrown.

jojochuang commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -669,6 +669,15 @@
+    <name>ozone.om.db.dirs.permissions</name>
+    <value>750</value>
+    <description>
+      Permissions for the metadata directories for Ozone Manager The
```
```suggestion
      Permissions for the metadata directories for Ozone Manager. The
```

ArafatKhan2198 replied:
Will make the change !! 
Thanks !!

devmadhuu reviewed (commented):
Thanks @ArafatKhan2198  for working on this patch. LGTM +1

ashishkumar50 commented:
Thanks for updating patch, LGTM.

jojochuang commented:
Merged. Thanks @ArafatKhan2198 for the PR and @ashishkumar50 and @devmadhuu for the review!
